File: 0494. Target Sum/dp.py
class Solution:
    def findTargetSumWays(self, nums: List[int], target: int) -> int:
        total = sum(nums)
        min, max = total * -1, total
        N = (2 * total + 1)
        prev, curr = [0] * N, [0] * N;
        prev[total] = 1;
        
        if(target < min or target > max):
            return 0;
        
        for num in nums:
            curr = [0] * N;
            for i in range(N):
                val = i - total
                
                if(val - num >= min):
                    idx = val - num + total;
                    curr[i] += prev[idx]
                    
                if(val + num <= max):
                    idx = val + num + total;
                    curr[i] += prev[idx]
            
            prev, curr = curr, prev
        
        return prev[target + total]

File: 0494. Target Sum/dp.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number}
 */
 var findTargetSumWays = function(nums, target) {
    const sum = nums.reduce((acc, num) => acc + num, 0);
    let min = -sum, max = sum, N = 2 * sum + 1; // sum = 10, [-10, -1], 0 ,[1, 10]
    let prev = [...Array(N)].fill(0), curr = [...Array(N)].fill(0);
    prev[sum] = 1;
    
    if(target > max || target < min) return 0;
    
    for(const num of nums){
        curr.fill(0);
        for(let i = 0; i < N; i++){
            const val = i - sum;
            
            if(val - num >= min){
                const idx = val - num + sum;
                curr[i] += prev[idx]                
            }
            
            if(val + num <= max){
                const idx = val + num + sum;
                curr[i] += prev[idx]
            }
        }
        
        [prev, curr] = [curr, prev]
    }
    
    
    return prev[target + sum];
};